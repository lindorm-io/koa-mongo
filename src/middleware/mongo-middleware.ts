import { DefaultState, Middleware } from "koa";
import { IKoaMongoContext } from "../types";
import { IMongoConnectionOptions, MongoConnection } from "@lindorm-io/mongo";

export const mongoMiddleware =
  (options: IMongoConnectionOptions): Middleware<DefaultState, IKoaMongoContext> =>
  async (ctx, next): Promise<void> => {
    const start = Date.now();

    ctx.client.mongo = new MongoConnection(options);

    await ctx.client.mongo.connect();

    ctx.logger.debug("mongo connection established");

    ctx.metrics.mongo = (ctx.metrics.mongo || 0) + (Date.now() - start);

    try {
      await next();
    } finally {
      await ctx.client.mongo.disconnect();
    }
  };
