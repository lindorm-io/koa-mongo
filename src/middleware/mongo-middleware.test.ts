import MockDate from "mockdate";
import { IMongoConnectionOptions, MongoConnection, MongoConnectionType } from "@lindorm-io/mongo";
import { mongoMiddleware } from "./mongo-middleware";
import { logger } from "../test";

MockDate.set("2020-01-01 08:00:00.000");

const next = jest.fn();

describe("mongoMiddleware", () => {
  let ctx: any;
  let inMemoryStore: Record<string, any>;
  let options: IMongoConnectionOptions;

  beforeEach(() => {
    inMemoryStore = { initialized: true };

    options = {
      type: MongoConnectionType.MEMORY,
      auth: {
        user: "user",
        password: "password",
      },
      url: {
        host: "host",
        port: 100,
      },
      databaseName: "databaseName",
      inMemoryStore,
    };

    ctx = { client: {}, logger, metrics: {} };
  });

  test("should set a mongo connection on context", async () => {
    await expect(mongoMiddleware(options)(ctx, next)).resolves.toBe(undefined);

    expect(ctx.client.mongo).toStrictEqual(expect.any(MongoConnection));

    const db = ctx.client.mongo.getDatabase();
    const collection = await db.collection("collectionName");
    await collection.createIndex({ index: "one" }, { options: "two" });

    expect(inMemoryStore).toMatchSnapshot();
  });
});
