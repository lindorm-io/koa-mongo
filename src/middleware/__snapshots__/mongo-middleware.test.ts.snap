// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mongoMiddleware should set a mongo connection on context 1`] = `
Object {
  "databaseName": Object {
    "collectionName": Object {
      "data": Array [],
      "indices": Array [
        Object {
          "index": Object {
            "index": "one",
          },
          "options": Object {
            "options": "two",
          },
        },
      ],
    },
  },
  "initialized": true,
}
`;
